octoprint_bambu_printer/printer/bambu_virtual_printer.py

```diff
@@ -580,6 +580,123 @@ def speed_adjust(speed_percentage):
+                                processed_color = color[:6] if color and len(color) >= 6 else "000000"
+
+                                parsed_ams_trays.append({
+                                    'global_id': global_id,
+                                    'type': material,
+                                    'color': processed_color,
+                                    'unit': unit_index,
+                                    'slot': slot_index
+                                })
+                                self._log.debug(f"REFINED_PARSE: Parsed tray Unit {unit_index}, Slot {slot_index}. Global ID: {global_id}, Type: {material}, Color: #{processed_color}")
+
+                            else:
+                                self._log.debug(f"REFINED_PARSE: Skipping malformed tray data (not dict or missing type/color) at Unit {unit_index}, Slot {slot_index}.")
+                    else:
+                        self._log.debug(f"REFINED_PARSE: Skipping malformed AMS unit container data (not dict or missing 'tray' list) at index {unit_index}.")
+
+            except Exception as e:
+                self._log.error(f"REFINED_PARSE: Error during cached data parsing: {type(e).__name__}: {e}", exc_info=True)
+                self.sendIO(f"echo: Error parsing cached AMS data: {e}\n")
+                parsed_ams_trays = [] # Clear list on error
+
+
+            # Now send the parsed data to the terminal
+            if parsed_ams_trays:
+                 for tray_data in parsed_ams_trays:
+                      output_line = (
+                           f"echo: AMS Unit {tray_data.get('unit', '?')}, Slot {tray_data.get('slot', '?')}"
+                           f" (Global ID: {tray_data.get('global_id', '?')})"
+                           f" - Type: {tray_data.get('type', 'Unknown')}, Color: #{tray_data.get('color', '000000')}\n" # Add # to color
+                      )
+                      self.sendIO(output_line)
+                      self._log.debug(f"REFINED_PARSE: Sent terminal output for Global ID {tray_data.get('global_id', '?')}.")
+            else:
+                 # This message will now be printed if the parsed_ams_trays list ends up empty
+                 # even if raw data was present, indicating a parsing failure.
+                 self.sendIO("echo: Parsed data list is empty (check raw data and logs for parsing errors).\n")
+
+
+            self.sendIO("echo: -----------------------------------------------------\n")
+
+        # Send the standard "ok N+1" response to acknowledge the command
+        try:
+            next_expected_line = self._serial_io.lastN + 1
+            self.sendIO(f"ok {next_expected_line}\n")
+        except Exception as e:
+             self._log.error(f"REFINED_PARSE: Error sending ok N+1: {type(e).__name__}: {e}", exc_info=True)
+
+        self._log.debug("M1111 command finished (refined parse).")
+        return True # Indicate the command was handled
+
+
```

Comment:
convert this to an @ command queueing phase and don't send by completing with `return None,`: https://docs.octoprint.org/en/master/plugins/hooks.html#octoprint-comm-protocol-atcommand-phase